### ISSUE TITLE
adding a pynvml wrapper

### ISSUE THREAD

**Issue by stas00** [1132719438/pynvx#2]
Thank you for adding this library, this is very useful, since now OSX users can be supported.

Would it be possible to add a wrapper so that `pynvx` can be used transparently by just loading it instead of `pynvml`, while utilizing `pynvml` syntax? Where the functionality overlaps of course.

For example, in our library we use only 4 functions, so I wrote this monkey patching wrapper hack:

```
import platform
import pynvml
from collections import namedtuple

use_gpu = torch.cuda.is_available()

GPUMemory = namedtuple('GPUMemory', ['total', 'free', 'used'])

is_osx = platform.system() == "Darwin"

# transparently monkey patch pynvx as pynvml API on OSX (for the few funcs we use)
if use_gpu and is_osx:
    try:
        import pynvx
    except:
        print("please install pynvx on OSX: pip install pynvx")
        sys.exit(1)

    # missing function
    def cudaDeviceGetHandleByIndex(id): return pynvx.cudaDeviceGetHandles()[id]
    setattr(pynvx, 'cudaDeviceGetHandleByIndex', cudaDeviceGetHandleByIndex)

    # different named and return value needs be a named tuple
    def cudaDeviceGetMemoryInfo(handle):
        info = pynvx.cudaGetMemInfo(handle)
        return GPUMemory(*info)
    setattr(pynvx, 'cudaDeviceGetMemoryInfo', cudaDeviceGetMemoryInfo)

    # remap the other functions
    for m in ['Init', 'DeviceGetCount', 'DeviceGetHandleByIndex', 'DeviceGetMemoryInfo']:
        setattr(pynvx, f'nvml{m}', getattr(pynvx, f'cuda{m}'))
    pynvml = pynvx
```
and then I can use it with the normal `pynvml` code, e.g.:
```
if use_gpu:
    pynvml.nvmlInit()
[...]
handle = pynvml.nvmlDeviceGetHandleByIndex(0)
info = pynvml.nvmlDeviceGetMemoryInfo(handle)
```

I only [tested it with these 4 functions that our library needed](https://github.com/fastai/fastai/commit/e6c7cc2001624f9c6e551426c89de9a12fbf4272). If it's not something you'd like to put work in, perhaps someone else would finish it up, so it'll be here as a starter.

Of course, another alternative, is to switch pynvx to mimic pynvml API so that it's a drop-in replacement. Either way works, it doesn't matter to me, as long as libraries needing to access nvml/cuda could have a transparent way to support all platforms and not each re-invent the wheel.

Thank you!

**Comment by 1132719438** [comment]
Nice idea, I will work on it soon.

**Comment by 1132719438** [comment]
I have added the pynvml wrapper implemention, please help to test.

Usage:

```
from pynvx import pynvml

try:
    pynvml.nvmlInit()
except NVMLError as e:
    print(e)

print('Device Count    : {}'.format(pynvml.nvmlDeviceGetCount()))
for index in range(pynvml.nvmlDeviceGetCount()):
    handle = pynvml.nvmlDeviceGetHandleByIndex(index)
    print('Device Name              : {}'.format(pynvml.nvmlDeviceGetName(handle)))
    mem_info = pynvml.nvmlDeviceGetMemoryInfo(handle)
    print('Device Memory            : {}, {}, {}'.format(mem_info.total, mem_info.used, mem_info.free))
pynvml.nvmlShutdown()
```

**Comment by stas00** [comment]
This is awesome, @1132719438!

It works well, tested both the above sample and in my app - on linux only.

Will integrate once you make a new release.

--------------------

Is it OK that your wrapper actually relies on pynvml being installed? I guess while pynvml doesn't work on OSX it does install and loads fine there, is that the assumption?

But in any case this distro then should either include these dependencies:

* python 3:
`nvidia-ml-py3` or `py3nvml` (2 versions with slightly different API I think)
* python 2:
`nvidia-ml-py`

or try/except and print a hint to install them.

https://github.com/fbcotter/py3nvml has extra API not present in `nvidia-ml-py` and its py3 port `nvidia-ml-py3` (https://github.com/nicolargo/nvidia-ml-py3)

-----------------------------

and btw, you might want to link to the python3 versions from your README.md, currently you link only to the python2 implementation.

---------------------------

and just found yet another py3 nvml bindings: 
https://github.com/SN9NV/python3-nvml

**Comment by stas00** [comment]
or and in `setup.py` this should be a list:

    keywords=["cuda", "python", "nvidia", "gpu", "macos", "OS X"],

**Comment by 1132719438** [comment]
Thanks @stas00 , version 1.0.0 has been released.

**Comment by stas00** [comment]
Fantastic, @1132719438. Thank you for this amazingly fast coding and release!

I ported https://github.com/stas00/ipyexperiments/ and https://github.com/fastai/fastai to use your wrapper instead of my hack. Tested on linux. Hoping it'll get tested soon on OSX by @phenomax.

Perhaps drop a note to https://github.com/nicolargo/nvidia-ml-py3 to link back to pynvx as a solution on OSX? 

wrt, `nvidia-ml-py` I'm not sure which one is being maintained, there are a few forks of it on github. probably no point wasting time there.
